### ISSUE TITLE
REGR: equality (__eq__) now ignores the z dimension

### ISSUE THREAD

**Issue by jorisvandenbossche** [shapely/shapely#1732]
With shapely 2.0, we switched the implementation of `Geometry.__eq__` to use the "equals_exact" predicate from GEOS, instead of manually comparing the coordinates as we did before (`geom1.coords == geom2.coords`), in the assumption this is equivalent. But it seems that `GEOSEqualsExact` only checks the X and Y coordinates (https://github.com/libgeos/geos/issues/809), thus introducing a regression in comparing 3D geometries:

With shapely 1.8
```python
>>> Point(1, 2, 3) == Point(1, 2, 4)
False
```

With shapely 2.0:

```python
>>> Point(1, 2, 3) == Point(1, 2, 4)
True
```

**Comment by idanmiara** [comment]
Hi @jorisvandenbossche Thanks for looking into this.
It would be useful to keep the current functionality (compare only xy) as a separate function, maybe `xy_equals(other)`.

**Comment by jorisvandenbossche** [comment]
That already exists as `equals_exact` (the naming doesn't make this clear, and we need to update this in the docs that it actually _is_ only for x and y)

**Comment by idanmiara** [comment]
> That already exists as `equals_exact` (the naming doesn't make this clear, and we need to update this in the docs that it actually _is_ only for x and y)

Thanks for the pointer! that reminds me of https://github.com/shapely/shapely/pull/1231 which I think almost ready (waiting for some final feedback there)

**Comment by idanmiara** [comment]
@jorisvandenbossche as I understand from https://github.com/libgeos/geos/issues/809#issuecomment-1405242815 this issue would be solved properly only in future GEOS version, right?
If so - I think we need a quick fix for `2.0.1` as this is a major regression. Should I suggest a quick PR fix for `2.0.1` or are you working on it already ?

**Comment by jorisvandenbossche** [comment]
Yes, we can't rely on that _right now_, so we still need to workaround for released versions of GEOS. I started on backporting https://github.com/libgeos/geos/pull/810 to our own code using the C API, and that could be used to fix this (although maybe a bit overkill for just this issue). Will open a PR shortly.
